### PR TITLE
RFQ - add ssm join capability

### PR DESCRIPTION
**Pull request by fatpelt**
There are obvious changes needing to be made here, but this is a request for comments on adding SSM to socket2.  this has been tested and validated on unix to function properly.

Signed-off-by: patrick felt <patrick.felt@sling.com>

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! Looks like there's some CI failures though?

File: src/sys/unix.rs [hunk 1]
```diff
@@ -67,6 +67,15 @@ pub const IPPROTO_TCP: i32 = libc::IPPROTO_TCP;
 pub const IPPROTO_UDP: i32 = libc::IPPROTO_UDP;
 pub const SOCK_SEQPACKET: i32 = libc::SOCK_SEQPACKET;
 pub const SOCK_RAW: i32 = libc::SOCK_RAW;
+pub const IP_ADD_SOURCE_MEMBERSHIP: libc::c_int = 39;
+
+#[repr(C, packed)]
+#[derive(Copy, Clone)]
+pub struct IpMreqSource {
```
Comment (sfackler) [inline src/sys/unix.rs]: These should be pushed up into the libc crate.

Comment (fatpelt) [reply src/sys/unix.rs]: yeah.  those i figured would have to upstream.  any other thoughts?

**Comment by fatpelt** [comment]
https://github.com/rust-lang/libc/pull/1578
https://github.com/retep998/winapi-rs/pull/829

**Comment by godsey** [comment]
I didn’t think of doing libc::c_int = like this. I’ve been working off a locally modified libc crate. The value here is different for MacOS and Windows. I’m stalled and somewhat lost trying to add the Filter by source and ICMP components ssm brings.

**Comment by fatpelt** [comment]
@godsey not sure i'm following your query?

**Comment by Thomasdezeeuw** [comment]
@fatpelt is this something you still want to add? If so, take a look at the new structure for `join_multicast_v4`: https://github.com/rust-lang/socket2/blob/819716c66f6c591388fb2ec117503f96f05f15d2/src/socket.rs#L796-L802.

Also are there also leave functions? And an option for IPv6?

**Comment by godsey** [comment]
> @godsey not sure i'm following your query?

Sorry for the late reply, I missed the question.

I ended up keeping a local_lib socket2 and added the following for SSM support:

    pub fn join_source_specific_multicast_v4(&self, multiaddr: &Ipv4Addr, sourceaddr: &Ipv4Addr, interface: &Ipv4Addr) -> io::Result<()> {
        let multiaddr = to_s_addr(multiaddr);
        let sourceaddr = to_s_addr(sourceaddr);
        let interface = to_s_addr(interface);
        let mreq_source = libc::ip_mreqn {
            imr_multiaddr: libc::in_addr { s_addr: multiaddr },
            imr_sourceaddr: libc::in_addr { s_addr: sourceaddr },
            imr_interface: libc::in_addr { s_addr: interface },
        };
        unsafe { self.setsockopt(libc::IPPROTO_IP, libc::IP_ADD_SOURCE_MEMBERSHIP, mreq_source) }
    }

    pub fn leave_source_specific_multicast_v4(&self, multiaddr: &Ipv4Addr, sourceaddr: &Ipv4Addr, interface: &Ipv4Addr) -> io::Result<()> {
        let multiaddr = to_s_addr(multiaddr);
        let sourceaddr = to_s_addr(sourceaddr);
        let interface = to_s_addr(interface);
        let mreq_source = libc::ip_mreqn {
            imr_multiaddr: libc::in_addr { s_addr: multiaddr },
            imr_sourceaddr: libc::in_addr { s_addr: sourceaddr },
            imr_interface: libc::in_addr { s_addr: interface },
        };
        unsafe { self.setsockopt(libc::IPPROTO_IP, libc::IP_DROP_SOURCE_MEMBERSHIP, mreq_source) }
    }


And then in the libc-rs src/unix/linux_like/mod.rs:

    pub struct ip_mreqn {
        pub imr_multiaddr: in_addr,
        pub imr_address: in_addr,
        pub imr_interface: in_addr,
    }

**Comment by Thomasdezeeuw** [comment]
@fatpelt due to age I'm closing this. If you're still interested in adding this please open a new pr.

**Pull request closed without merging**